### PR TITLE
Gitcoin claimer service

### PR DESCRIPTION
**Pull request by liringlas**
https://www.notion.so/Gitcoin-for-Faucet-8be81dff9c5744feae8430400ed12ee3

### CODE REVIEW

File: faucet-client/src/common/FaucetConfig.ts [hunk 1]
```diff
@@ -15,6 +15,7 @@ export interface IFaucetConfig {
     pow?: IPoWModuleConfig;
     passport?: IPassportModuleConfig;
   };
+  gitcoinMinimumScore?: number;
```
Comment (annielinnik) [inline faucet-client/src/common/FaucetConfig.ts:18]: Shouldn't 'gitcoinApiToken' and 'gitcoinScorerId' be declared here?

File: src/db/FaucetDatabase.ts [hunk 2]
```diff
@@ -251,49 +371,73 @@ export class FaucetDatabase {
...
+    );
   }
 
+  private selectGitcoinClaims(
+    whereSql: string,
+    whereArgs: any[]
+  ): Promise<GitcoinClaimTableType[]> {
+    return this.db.all(
+      "SELECT " +
+        GitcoinClaimsColumns.join(",") +
```
Comment (annielinnik) [inline src/db/FaucetDatabase.ts]: Why don't you use '*'?

File: src/eth/EthWalletManager.ts [hunk 3]
```diff
@@ -67,8 +70,10 @@ export class EthWalletManager {
   private txReceiptPollInterval = 12000;
 
   public async initialize(): Promise<void> {
-    if(this.initialized)
+    if(this.initialized) {
+      console.log("EthWalletManager already initialized");
```
Comment (annielinnik) [inline src/eth/EthWalletManager.ts]: ServiceManager.GetService(FaucetProcess).emitLog

File: src/eth/EthWalletManager.ts [hunk 4]
```diff
@@ -328,6 +333,7 @@ export class EthWalletManager {
     let txPromise: Promise<TransactionReceipt>;
     let retryCount = 0;
     let txError: Error;
+
     const buildTx = () => {
```
Comment (annielinnik) [inline src/eth/EthWalletManager.ts]: this.buildTx?

File: src/eth/EthWalletManager.ts [hunk 5]
```diff
@@ -384,7 +390,72 @@ export class EthWalletManager {
     };
   }
 
-  public async sendCustomTx(params: {target: string, amount: bigint, data?: string, gasLimit: number}): Promise<TransactionResult> {
+  private async buildTx(target: string, amount: number | string) {
+    const nonce = this.walletState.nonce;
```
Comment (annielinnik) [inline src/eth/EthWalletManager.ts:365]: Are there any other differences from `sendClaimTx` except for this line? 
How about calling this.sendClaimTx here?

File: src/webserv/FaucetWebApi.ts [hunk 6]
```diff
@@ -213,137 +330,237 @@ export class FaucetWebApi {
...
+        (session) =>
+          session.getSessionStatus() === FaucetSessionStatus.RUNNING &&
+          session.getUserId() === userId
+      );
+    if (activeSessions.length > 0) {
+      throw new FaucetError(
+        "GITCOIN_CLAIM_ERROR",
+        "Please stop your active mining session before claiming Gitcoin rewards."
+      );
+    }
```
Comment (annielinnik) [inline src/webserv/FaucetWebApi.ts]: Isn't it the same as checkIfUserCanClaimGitcoin() ?

File: src/webserv/FaucetWebApi.ts [hunk 7]
```diff
@@ -213,137 +330,237 @@ export class FaucetWebApi {
...
           failedReason: ex.toString(),
-        }
+        };
       }
     }
 
     return sessionInfo;
   }
 
+  private async onClaimRewardByGitcoin(
```
Comment (annielinnik) [inline src/webserv/FaucetWebApi.ts:476]: What if another gitcoin claim from this user is in progress, but not finished yet?

**Review by annielinnik — CHANGES_REQUESTED** [review]

File: src/eth/EthWalletManager.ts (hunk 4 above)
Comment (liringlas) [reply src/eth/EthWalletManager.ts]: Yeah I used it here too, then encountered some errors, reverted it.
Then figured that the error was Web3 related.
I'll fix it here and use `this.buildTx()` too!

File: src/eth/EthWalletManager.ts (hunk 5 above)
Comment (liringlas) [reply src/eth/EthWalletManager.ts:365]: If I understand your question correctly:
1) When using `sendClaimTx` it mutates `EthClaimInfo` param (with PoW session-related info) that is passed there, so then later that mutated param is being used in other functions. 
So I decided to use custom `sendGitcoinClaimTx`. Maybe later we can make it to use unified method function, but right now I thought to leave `sendClaimTx` as it is.

File: src/webserv/FaucetWebApi.ts (hunk 7 above)
Comment (liringlas) [reply src/webserv/FaucetWebApi.ts:476]: Good catch, will fix.

File: faucet-client/src/common/FaucetConfig.ts (hunk 1 above)
Comment (liringlas) [reply faucet-client/src/common/FaucetConfig.ts:18]: No, it's client-side facing config, no need for API token and scorerId there.

File: src/webserv/FaucetWebApi.ts (hunk 6 above)
Comment (liringlas) [reply src/webserv/FaucetWebApi.ts]: Thanks for noticing

**Review by annielinnik — APPROVED** [review]

**Pull request merged**